src/Controller/PrepaidBalanceController.php

```diff
@@ -0,0 +1,289 @@
+      });
+
+      $billing_documents = $this->getDataFromCache($user, 'billing_documents', function () {
+        return $this->monetization->getBillingDocumentsMonths();
+      });
+
+      // Build the form.
+      $build['download_form'] = $this->formBuilder->getForm(PrepaidBalanceReportsDownloadForm::class, $user, $supported_currencies, $billing_documents);
+      $build['download_form']['#cache']['keys'] = [$this->getCacheId($user, 'download_form')];
+    }
+
+    return $build;
+  }
+
+  /**
+   * Returns the cache max age.
+   *
+   * @return int
+   *   The cache max age.
+   */
+  protected function getCacheMaxAge() {
+    // Get the max-age from config.
+    if ($config = $this->config(BillingConfigForm::CONFIG_NAME)) {
+      return $config->get('prepaid_balance.cache_max_age');
+    }
+
+    return 0;
+  }
+
+  /**
+   * Helper to check if user has access to refresh prepaid balance.
+   *
+   * @param \Drupal\user\UserInterface $user
+   *   The user entity.
+   *
+   * @return \Drupal\Core\Access\AccessResult
+   *   If $condition is TRUE, isAllowed() will be TRUE, otherwise isNeutral()
+   *   will be TRUE.
+   */
+  protected function canRefreshBalance(UserInterface $user) {
+    return $this->currentUser->hasPermission('refresh any prepaid balance') ||
+      ($this->currentUser->hasPermission('refresh own prepaid balance') && $this->currentUser->id() === $user->id());
+
+    // @TODO Figure out why AccessResult is not working her.
+    return AccessResult::allowedIf(
+      $this->currentUser->hasPermission('refresh any prepaid balance') ||
+      ($this->currentUser->hasPermission('refresh own prepaid balance') && $this->currentUser->id() === $user->id())
+    )->cachePerPermissions();
+  }
+
+  /**
+   * Helper to retrieve data from cache.
+   *
+   * @param \Drupal\user\UserInterface $user
+   *   The user entity.
+   * @param string $suffix
+   *   The cache id suffix.
+   * @param callable $callback
+   *   The callback if not in cache.
+   *
+   * @return mixed
+   *   The data.
+   */
+  protected function getDataFromCache(UserInterface $user, string $suffix, callable $callback) {
+    $max_age = $this->getCacheMaxAge();
+
+    // If caching is disable, run callback and return.
+    if ($max_age == 0) {
+      return $callback();
+    }
+
+    $cid = $this->getCacheId($user, $suffix);
+
+    // Check cache.
+    if ($cache = $this->cache()->get($cid)) {
+      return $cache->data;
+    }
+
+    $data = $callback();
+    $this->cache()
+      ->set($cid, $data, time() + $max_age, $this->getCacheTags($user));
+
+    return $data;
+  }
+
+  /**
+   * Helper to get the billing cache tags.
+   *
+   * @param \Drupal\user\UserInterface $user
+   *   The user entity.
+   *
+   * @return array
+   *   The cache tags.
+   */
+  public static function getCacheTags(UserInterface $user) {
+    return [
+      static::CACHE_PREFIX,
+      static::CACHE_PREFIX . ':user:' . $user->id(),
```

Comment:
Can `static::CACHE_PREFIX . ':user:' . $user->id()` use `static::getCacheId()` to keep the id building all in one place?

How is `static::CACHE_PREFIX` used? I'm wondering if/why we need more that just `'apigee.monetization.billing:user:' . $user_id`?

Should `apigee.monetization.billing` be changed to `apigee.monetization.prepaid_balance`?

Reply:
The `static::CACHE_PREFIX` tag is used to clear all cache when the max_age is updated. See submit handler for the config form.

I've renamed billing to prepaid balance (I was planning to do this in another PR)